shop.cpp

```diff
@@ -1,26 +1,77 @@
 #include "shop.h"
-
+#include <chrono>
 
 Shop::Shop(int mMaxChairs)
     :mMaxChairs(mMaxChairs)
-    ,mTheBarber(this)
-    ,mBarberSemaphore(0)
+    ,mBarberSemaphore(1)
     ,mCustomersSemaphore(0)
 {
+    mTheBarber = new Barber(this);
+}
+
+Shop::~Shop()
+{
+    delete mTheBarber;
+    mTheCustomers.clear();
+}
+
+void Shop::newCustomerArrives(Customer *newCustomer)
+{
+    std::lock_guard<std::mutex> lock(mShopMutex);
+    mTheCustomers.emplace_back(newCustomer);
+}
+
+bool Shop::emptySeatsExist()
+{
+    std::lock_guard<std::mutex> lock(mShopMutex);
+    return ((mWaitingCustomers < mMaxChairs));
+}
+
+void Shop::addWaitingCustomer()
+{
+    std::lock_guard<std::mutex> lock(mShopMutex);
+    mWaitingCustomers++;
+}
 
+void Shop::removeWaitingCustomer()
+{
+    std::lock_guard<std::mutex> lock(mShopMutex);
+    mWaitingCustomers--;
+    clearFirstTerminatedCustomerFound();
 }
 
-void Shop::newCustomer(std::string customerName)
+void Shop::clearFirstTerminatedCustomerFound()
 {
-    mCustomersMapMutex.lock();
-    mTheCustomers.emplace(customerName, Customer(customerName, this));
-    mCustomersMapMutex.unlock();
+    std::lock_guard<std::mutex> lock(mShopMutex);
+    // Check if Customers have been finished
+    auto it = mTheCustomers.begin();
+    Customer* c;
+    for (;it != mTheCustomers.end(); ++it)
+    {
+        if (!((*it)->isTerminated()))
+        {
+            continue;
+        }
+        c = (*it);
+        break;
+    }
+    *it = mTheCustomers.back();
```

Comment:
Are you trying here to remove the deleted customer from the list? I think it will not work.

Reply:
Yes, this is my purspose. I will simplify it as I mentioned above. :ok: